nodeJS/authentication/session_based_authentication.md

```diff
@@ -0,0 +1,437 @@
+
+### Storing passwords securely
+
+The most secure way to store passwords? Don't. Offloading that responsibility to other systems by letting users log in with their Facebook or Google accounts means you won't need to store passwords on your side at all. That being said, that's very much out of scope right now and doesn't really help us with learning these fundamentals behind the scenes.
+
+By far the worst way we can store passwords is to just store them in plaintext like we've done in our example app earlier. Even if we encrypted the passwords, all an attacker would need is the key to decrypt all the passwords. Let's face it, if someone managed to gain access to your database, it probably wouldn't be very hard for them to get the encryption key (assuming they don't already have it).
+
+Remember [hash functions](https://www.theodinproject.com/lessons/javascript-hashmap-data-structure#what-is-a-hash-code) from the Hashmap lesson? We want to hash our passwords, then store the hash since hashes are one-way functions. We also want to [salt](https://en.wikipedia.org/wiki/Salt_(cryptography)) the password when hashing to prevent identical passwords from being stored with identical hashes. On top of all that, we also want the hash function to be purposely slow - not so slow that a normal user will be waiting ages just to log in but certainly slow enough to minimize the number of attempts an attacker might be able to make in a given amount of time.
+
+#### Argon2
+
+All of this can be done for us via the [argon2 npm package](https://www.npmjs.com/package/argon2). Fortunately for us, using it doesn't require that much extra work. Going back to our `POST /signup` middleware, let's hash our password before we store it:
+
+```bash
+npm install argon2
+```
+
+```javascript
+const argon2 = require("argon2");
+
+app.post("/signup", async (req, res, next) => {
+  try {
+    const hashedPassword = await argon2.hash(req.body.password);
+    await pool.query(
+      "INSERT INTO users (username, password) VALUES ($1, $2)",
+      [req.body.username, hashedPassword],
+    );
+    res.redirect("/");
+  } catch(err) {
+    next(err);
+  }
+});
+```
+
+We don't need to modify any of its options, as the defaults all meet the [password storage recommendations set by OWASP](https://cheatsheetseries.owasp.org/cheatsheets/Password_Storage_Cheat_Sheet.html#introduction) (Open Worldwide Application Security Project). Now in our `POST /login` middleware, we can also use argon2 to verify the submitted password against the stored salted hash:
+
+```javascript
+app.post("/login", async (req, res, next) => {
+  try {
+    const { rows } = await pool.query(
+      "SELECT * FROM users WHERE username = $1",
+      [req.body.username],
+    );
+    const user = rows[0];
+
+    // argon2.verify requires an argon2 hash as the first argument
+    // so we must early return if there is no matching user
```

Comment:
There's a timing attack here, which leaks usernames. Do we think a note about this? Or just ignore it

Reply:
I'm surprised I glossed over this, considering I mention generic validation feedback earlier in the lesson. Will address this - thinking of using a predetermined argon2id hash as a fallback if `user?.password` short circuits so that `argon2.verify()` can still run fine. Then only go down the happy path if both user exists and password matches.

Any alternative suggestions?

Reply:
That's probably the simplest approach. You probably want to pick a huge random string to hash, since if you use something that someone actually tries as a password, you might accidentally grant a session to a user that doesn't exist. Maybe hash a uuid.
Obligatory I am not a cybersec guy message

Reply:
Yeah, the random argon2 hash would just be to force the hash verification. Then change the following conditional to `if (user && isMatchingPassword)`